File: data-structure-algo-directory/Numerical_method/C++/bisection.cpp
/*
Bisection Method to find the root of polynomial function
*/

#include <iostream>
#include <cmath>
#include <iomanip>
using namespace std;

// Define the polynomial function to find the root.
double f(double x) {
    return x * x - 4 * x - 10;
}

double bisection(double a, double b, double tol) {
    if (f(a) * f(b) >= 0) {
        cout << "Bisection method fails. The function must have different signs at the endpoints a and b." << endl;
        return -1;
    }
    
    double c = a;
    int iteration = 0;

    // Set precision to 3 decimal places
    cout << fixed << setprecision(3);
    
    cout << setw(10) << "Iteration" << setw(15) << "a" << setw(15) << "b" << setw(15) << "c" << setw(20) << "f(c)" << endl;
    cout << "----------------------------------------------------------------------------" << endl;

    while ((b - a) >= tol) {
        c = (a + b) / 2; // Find the midpoint
        iteration++;

        cout << setw(10) << iteration << setw(15) << a << setw(15) << b << setw(15) << c << setw(20) << f(c) << endl;
        
        // Check if the midpoint is a root
        if (f(c) == 0.0) {
            break;
        }
        // Decide the side to repeat the steps
        else if (f(c) * f(a) < 0) {
            b = c;
        }
        else {
            a = c;
        }
    }
    return c;
}

int main() {
    double a, b, tol;
    
    cout << "Enter the left endpoint a: ";
    cin >> a;
    cout << "Enter the right endpoint b: ";
    cin >> b;
    cout << "Enter the tolerance: ";
    cin >> tol;
    
    double result = bisection(a, b, tol);
    if (result != -1) {
        cout << "----------------------------------------------------------------------------" << endl;
        // Reset precision to default after printing the table
        cout << fixed << setprecision(6);
        cout << "The root of the polynomial within the given tolerance is: " << result << endl;
    }

    return 0;
}

File: data-structure-algo-directory/Numerical_method/Python/bisection.py
"""
Bisection Method to find the root of polynomial function
"""
def f(x):
    # Define the polynomial function to find the root.
    #Hardcoding for eqn :)
    return x * x - 4*x - 10

def bisection(a, b, tol):
    if f(a) * f(b) >= 0:
        print("Bisection method fails. The function must have different signs at the endpoints a and b.")
        return None
    
    print(f"{'Iteration':>10} {'a':>15} {'b':>15} {'c':>15} {'f(c)':>20}")
    print("-" * 82)
    
    c = a
    iteration = 0

    while (b - a) >= tol:
        c = (a + b) / 2  # Find the midpoint
        iteration += 1

        print(f"{iteration:>10} {a:>15.3f} {b:>15.3f} {c:>15.3f} {f(c):>20.3f}")
        
        # Check if the midpoint is a root
        if f(c) == 0.0:
            break
        # Decide the side to repeat the steps
        elif f(c) * f(a) < 0:
            b = c
        else:
            a = c
    return c

def main():
    a = float(input("Enter the left endpoint a: "))
    b = float(input("Enter the right endpoint b: "))
    tol = float(input("Enter the tolerance: "))
    
    result = bisection(a, b, tol)
    if result is not None:
        print("-" * 82)
        print(f"The root of the polynomial within the given tolerance is: {result:.3f}")

if __name__ == "__main__":
    main()
